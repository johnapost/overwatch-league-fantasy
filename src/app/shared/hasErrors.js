// @flow

export default (fieldsError: Object) =>
  Object.keys(fieldsError).some(field => fieldsError[field]);
