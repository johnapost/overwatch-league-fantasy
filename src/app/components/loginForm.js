// @flow
/* eslint-disable jsx-a11y/anchor-is-valid */

import React, { Component } from "react";
import { compose } from "redux";
import { isLoaded } from "react-redux-firebase";
import { Form, Button, Input, message } from "antd";
import { withRouter } from "next/router";
import withFirestore from "../shared/withFirestore";
import hasErrors from "../shared/hasErrors";

type Props = {
  disabled: boolean,
  firebase: Object,
  form: Object,
  router: Object,
  setDisabled: (bool: boolean) => void,
  showSignUpModal: Function
};

export class LoginFormComponent extends Component<Props> {
  emailEl: ?HTMLElement = null;

  componentDidMount() {
    if (this.emailEl) {
      this.emailEl.focus();
    }
  }

  handleSubmit = (e: SyntheticEvent<*>) => {
    e.preventDefault();
    const {
      firebase,
      form: { validateFields },
      setDisabled
    } = this.props;

    if (!isLoaded(firebase)) {
      return;
    }

    setDisabled(true);
    const closeMessage = message.loading("Logging in..", 0);

    const finishSubmitting = () => {
      setDisabled(false);
      closeMessage();
    };

    validateFields((err, { email, password }) => {
      if (err) {
        finishSubmitting();
        return;
      }

      firebase
        .auth()
        .signInWithEmailAndPassword(email, password)
        // Successful login
        .then(finishSubmitting)
        .catch(({ code, message: errorMessage }) => {
          finishSubmitting();
          message.error(errorMessage);
          // eslint-disable-next-line no-console
          console.error(code, errorMessage);
        });
    });
  };

  render() {
    const {
      disabled,
      form: { getFieldDecorator, getFieldsError },
      router: {
        query: { email }
      },
      showSignUpModal
    } = this.props;
    const emailValidations = [{ required: true }, { type: "email" }];
    const passwordValidations = [{ required: true }];

    return (
      <Form layout="inline" onSubmit={this.handleSubmit}>
        <Form.Item help={false}>
          {getFieldDecorator("email", {
            rules: emailValidations,
            initialValue: email
          })(
            <Input
              placeholder="Email"
              ref={el => {
                this.emailEl = el;
              }}
            />
          )}
        </Form.Item>
        <Form.Item help={false}>
          {getFieldDecorator("password", { rules: passwordValidations })(
            <Input placeholder="Password" type="password" />
          )}
        </Form.Item>
        <Form.Item help={false}>
          <Button
            type="primary"
            htmlType="submit"
            disabled={disabled || hasErrors(getFieldsError())}
          >
            Login
          </Button>
        </Form.Item>
        <Form.Item help={false}>
          <Button type="secondary" onClick={showSignUpModal}>
            Sign up
          </Button>
        </Form.Item>
      </Form>
    );
  }
}

export default compose(
  withRouter,
  withFirestore(),
  Form.create()
)(LoginFormComponent);
