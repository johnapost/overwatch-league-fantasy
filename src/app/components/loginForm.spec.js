import React from "react";
import { shallow } from "enzyme";
import { LoginFormComponent } from "./loginForm";
import hasErrors from "../shared/hasErrors";

jest.mock("../shared/hasErrors");

describe("LoginForm", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    hasErrors.mockImplementation(() => false);
  });

  describe("component", () => {
    const props = {
      disabled: false,
      form: {
        getFieldDecorator: jest.fn(() => jest.fn(component => component)),
        getFieldsError: jest.fn()
      },
      router: { query: { email: "fdsa@asdf.com" } }
    };

    it("should render fine", () => {
      const wrapper = shallow(<LoginFormComponent {...props} />);
      expect(wrapper.html()).toMatchSnapshot();
    });
  });
});
