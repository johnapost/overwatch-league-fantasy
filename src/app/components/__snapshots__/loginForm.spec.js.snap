// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm component should render fine 1`] = `"<form class=\\"ant-form ant-form-inline\\"><div class=\\"ant-row ant-form-item\\"><div class=\\"ant-form-item-control-wrapper\\"><div class=\\"ant-form-item-control\\"><span class=\\"ant-form-item-children\\"><input type=\\"text\\" placeholder=\\"Email\\" value=\\"\\" class=\\"ant-input\\"/></span></div></div></div><div class=\\"ant-row ant-form-item\\"><div class=\\"ant-form-item-control-wrapper\\"><div class=\\"ant-form-item-control\\"><span class=\\"ant-form-item-children\\"><input type=\\"password\\" placeholder=\\"Password\\" value=\\"\\" class=\\"ant-input\\"/></span></div></div></div><div class=\\"ant-row ant-form-item\\"><div class=\\"ant-form-item-control-wrapper\\"><div class=\\"ant-form-item-control\\"><span class=\\"ant-form-item-children\\"><button type=\\"submit\\" class=\\"ant-btn ant-btn-primary\\"><span>Login</span></button></span></div></div></div><div class=\\"ant-row ant-form-item\\"><div class=\\"ant-form-item-control-wrapper\\"><div class=\\"ant-form-item-control\\"><span class=\\"ant-form-item-children\\"><button type=\\"button\\" class=\\"ant-btn ant-btn-secondary\\"><span>Sign up</span></button></span></div></div></div></form>"`;
